config/initializers/passport.js

```diff
@@ -29,5 +38,225 @@ export function init(passport) {
     }
   }));
+
+  // Facebook
+  if (!isEmpty(oauthConfig.facebookClientId)) {
+    passport.use(new FacebookStrategy(
+      {
+        clientID:          oauthConfig.facebookClientId,
+        clientSecret:      oauthConfig.facebookClientSecret,
+        callbackURL:       `${config.host}/v2/oauth/facebook/callback`,
+        profileFields:     ['id', 'displayName', 'name', 'profileUrl', 'emails', 'photos'],
+        passReqToCallback: true,
+        enableProof:       true,
+      },
+      getStrategyCallback('facebook')
+    ));
+  }
+
+  // Google
+  // Enable Google+ api in app settings.
+  if (!isEmpty(oauthConfig.googleClientId)) {
+    passport.use(new GoogleStrategy(
+      {
+        clientID:          oauthConfig.googleClientId,
+        clientSecret:      oauthConfig.googleClientSecret,
+        callbackURL:       `${config.host}/v2/oauth/google/callback`,
+        passReqToCallback: true,
+      },
+      getStrategyCallback('google')
+    ));
+  }
+
+  // Github
+  if (!isEmpty(oauthConfig.githubClientId)) {
+    passport.use(new GithubStrategy(
+      {
+        clientID:          oauthConfig.githubClientId,
+        clientSecret:      oauthConfig.githubClientSecret,
+        callbackURL:       `${config.host}/v2/oauth/github/callback`,
+        passReqToCallback: true,
+      },
+      getStrategyCallback('github')
+    ));
+  }
+}
+
+/**
+ * Example:
+ *  john.doe@gmail.com => johndoe{n}
+ *  John Doe => JohnDoe{n}
+ */
+export async function generateUsername(dbAdapter, { firstName, lastName, email, username }) {
+  async function isUsernameInvalid(username) {
+    const user = await dbAdapter.getUserByUsername(username);
+
+    if (
+      user !== null ||
+      User.stopList().includes(username)
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  let generatedUsername;
+  if (isString(username)) {
+    generatedUsername = username;
+  } else if (isString(email)) {
+    // eslint-disable-next-line prefer-destructuring
+    generatedUsername = email.split('@')[0];
+  } else if (isString(firstName) && isString(lastName)) {
+    generatedUsername = `${firstName}${lastName}`;
+  } else {
+    throw new Error(`Could not generate username`);
+  }
+
+  // Filter out all non-alphanumeric symbols
+  generatedUsername = generatedUsername.replace(/\W/gi, '');
+
+  // eslint-disable-next-line no-await-in-loop
+  for (let n = 1, old = generatedUsername; await isUsernameInvalid(generatedUsername); ++n) {
+    generatedUsername = `${old}${n}`;
+  }
+
+  return generatedUsername;
+}
+
+/**
+ * Returns HTML with a script that calls window.oauthCallback of the parent window and closes the popup.
+ */
+export function renderCallbackResponse({ error, authToken }) {
+  return (`<!DOCTYPE html>
+<html>
+  <head>
+    <script>
+      var error = ${JSON.stringify(error && error.message)};
+      var authToken = ${JSON.stringify(authToken)};
+      window.opener.postMessage({ error: error, authToken: authToken }, '*');
```

Comment:
Is it safe to use '*' here?

Reply:
Yeah, this is my oversight. Should have replaced it with host.

Reply:
[https://github.com/FreeFeed/freefeed-ansible/blob/6bbcad420c5bdcf167c6e8ad08c8cea3b2c6d52f/roles/pepyatka/defaults/main.yml#L11](In the real world) the config.origin is actually a '*':) So may be we need to get origin from client before the start of authorization process?

Reply:
Done. I used cookies to temporarily store the origin.